torch/csrc/autograd/python_variable.cpp

```diff
@@ -2456,4 +2487,21 @@ c10::Layout concrete_layout_fn(
   return toLayout(out.ptr());
 }
 
+template<const char* func_name, typename... Ts>
+void concrete_trace_cuda(const c10::impl::PyInterpreter*, Ts... args) {
+  pybind11::gil_scoped_acquire gil;
+  at::impl::MaybeSetTLSOnEntryGuard guard;
+
+  try {
+    py::module mod = py::module::import("torch.utils._cuda_trace");
+    py::object hook = mod.attr(func_name).attr("fire_callbacks");
+    hook(args...);
+  }
+  catch (const py::error_already_set& e) {}
+  // The error is silently caught, because in most cases the import will fail
+  // when objects are destructed at the end of the program. This is expected,
+  // but unfortunately there is no way to tell if the import failed because
+  // the program has ended or because of another, unexpected reason.
```

Comment:
Hmm... this is a new problem. I wonder if there is a cleaner way to avoid this.

Most of the preexisting PyInterpreter hooks never get triggered during cleanup, since they're all about serving tensor operations that don't occur during shutdown. But the CUDA allocator events are different, since they are typically triggered upon tensor destruction so they do get served at shutdown.

I feel like what I would like is to know if the Python interpreter is shutting down, and if it is, stop serving these requests. Sometimes we've used `Py_IsInitialized()` but I don't know if that's sufficient.

Reply:
I've tried it in the code and I haven't gotten a segfault, after the interpreter was shut down `Py_IsInitialized` was equal to `false`. So it seems to work in this case.